lib/experimental/graph.nim

```diff
@@ -0,0 +1,428 @@
+
+  ## Perform breadth-first iteration of parent graph, starting from `node`
+  var que = initDeque[Id]()
+  que.addLast(root)
+  var visited: HashSet[Id]
+  while 0 < len(que):
+    let top = que.popFirst()
+    yield top
+    for (src, target) in graph.outEdges(top):
+      if target notin visited:
+        visited.incl target
+        que.addLast(target)
+
+
+proc getNoIncoming*[Id](graph: HGraph[Id]): seq[Id] =
+  ## Get all nodes with zero in degree of zero
+  let degrees = computeInDegrees(graph)
+  for node in graph.nodes:
+    if degrees[node] == 0:
+      result.add node
+
+proc topologicalOrdering*[Id](
+    graph: HGraph[Id], start: seq[Id] = @[]): seq[Id] =
+  ## Return graph nodes in topological ordering if possible. Otherwise (if
+  ## graph contains cycles raise `HGraphCyclesError`)
+  var active = toHashSet(start)
+  if active.len() == 0:
+    active = toHashSet(getNoIncoming(graph))
+
+  var graph = graph
+
+  while 0 < len(active):
+    let node = active.pop
+    result.add node
+    for (src, target) in graph.outEdges(node):
+      graph.removeEdge((src, target))
+      if graph.inDeg(target) == 0:
+        active.incl target
+
+  if 0 < graph.edgeCount:
+    raise HGraphCyclesError(
+      msg: "Topological sorting is impossible, graph contains cycles")
+
+
+template findMaxIt*(s: untyped, op: untyped): untyped =
+  var res: int = 0
+  var idx = 0
+  var prev: typeof((let it {.inject.} = s[0]; op))
+
+  for it {.inject.} in s:
+    let val = op
+    if val > prev:
+      res = idx
+
+    inc idx
+
+  res
+
+template findIt*(s: typed, op: untyped): int =
+  ## Find first element of the sequence for which `op` evaluates as true
+  ## and return it's index. If no such element is found return `-1`
+
+  var result = -1
+  for idx, it {.inject.} in s:
+    if op: result = idx; break
+
+  result
+
+
+proc graphEasyRepr*[Id](
+    graph: HGraph[Id], name: proc(id: Id): string): string =
+
+  for (src, dst) in edgePairs(graph):
+    result.add "[ $# ] $# [ $# ]\n" % [
+      name(src),
+      "->",
+      name(dst)
+    ]
+
+proc findCycles*[Id](
+    graph: HGraph[Id],
+    ignoreSelf: bool = false,
+    overrideDirected: bool = false,
+    yieldMultiedge: bool = false
+  ): seq[seq[Id]] =
+
+  var
+    visited: HashSet[Id]
+    stack: seq[Id]
+    resCycles: seq[seq[Id]]
+
+  proc printCycle(vertex: Id) =
+    var stack2: seq[Id]
+    stack2.add(stack.pop())
+
+    while stack2[^1] != vertex:
+      stack2.add(stack.pop())
+
+    var path: seq[Id]
+    while 0 < len(stack2):
+      path.add stack2[^1]
+      stack.add(stack2.pop())
+
+    resCycles.add path
+
+  proc processDfs() =
+    let top = stack[^1]
+    template body(vertex: untyped): untyped {.dirty.} =
+      if vertex == top and ignoreSelf:
+        discard
+
+      elif vertex in stack:
+        printCycle(vertex)
+
+      elif vertex notin visited:
+        stack.add vertex
+        processDFS()
+
+
+    for (_,vertex) in outEdges(graph, top):
```

Comment:
```suggestion
    let top = stack[^1]

    template body(vertex: untyped): untyped {.dirty.} =
      if vertex == top and ignoreSelf:
        discard
      elif vertex in stack:
        printCycle(vertex)
      elif vertex notin visited:
        stack.add vertex
        processDFS()

    for (_,vertex) in outEdges(graph, top):
```

The idea with the above is the space between the let and template declaration interrupt reading flow to emphasize we're changing gears somewhat and defining a helper. Keeping the if/elif branches tightly packed indicates this is one logical unit of templated code. When a break (blank line) is see it's clear we've returned back to our regularly scheduled program of proc declaration. Two new lines as in other parts of this file typically hint at breaking up whole routines. We don't have braces so all this subtlety is in the spacing, fun!